## 재검증 — 8c56081124 (읽기 전용, 절대경로)

주의: 이번 재검증은 `.worktrees/g6-usage-semantics`가 여전히 87e0f90 기준으로 체크아웃된 상태에서 진행했습니다. 지시받은 읽기 전용 제약상 이 워크트리를 새 커밋으로 checkout하거나 별도 worktree를 새로 만들지 않았습니다 — 대신 `git fetch` + `git show`/`git diff <sha>..<sha>`로 8c56081124의 실제 코드를 직접 읽었고, 로컬 재빌드/재실행 대신 **이 커밋에 대해 새로 트리거된 CI 런(31959938902)**을 1차 증거로 썼습니다. 이 제약 때문에 로컬 `dune build`/`runtest`는 이번엔 직접 재실행하지 못했다는 점을 먼저 밝힙니다.

---

### P1-1 (Claude Code CLI 경로) — **코드는 정확, 그러나 CI 게이트 FAIL로 아직 PASS 아님**

**반증 시도**: diff를 라인 단위로 읽고 (a) 매핑이 진짜 `usage_of_wire_counts`를 호출하는지, (b) absent→0 관례가 기존 패턴과 일치하는지, (c) 테스트가 계약을 고정하는지 확인.

**결과**:
- (a) `keeper_claude_code_runtime.ml`이 `Agent_core.Llm_provider.Backend_anthropic.usage_of_wire_counts ~input_tokens:usage.input_tokens ~output_tokens:usage.output_tokens ~cache_creation_input_tokens:usage.cache_creation_input_tokens ~cache_read_input_tokens:usage.cache_read_input_tokens`를 실제로 호출합니다. 가짜/그림자 재구현이 아니라 진짜 공유 생성자 경유. ✅
- (b) `Option.value (int_field "cache_creation_input_tokens") ~default:0` 패턴은 제가 1차 리뷰에서 읽은 `streaming.ml`의 `to_int_option |> Option.value ~default:0`과 동일 관례입니다. "다른 Anthropic-format 파스와 동일 관례"라는 주장은 사실입니다. ✅
- (c) `test_result_usage_is_carried`가 프레임 파서(`Runtime_claude_code.parse_result`) 출력에 `cache_read_input_tokens=42`(픽스처 보존), `cache_creation_input_tokens=0`(부재)을 고정합니다. 다만 이건 **프레임 파서 레벨**만 고정하고, keeper 최종 `Agent_core.Types.api_usage.input_tokens`가 실제로 `123456+42+0`이 되는지는 통합 테스트가 없습니다 — 팀장이 자인한 대로 6줄 필드 릴레이라 리스크는 낮지만, 정확히는 "미검증"입니다(nit, 블로킹 아님).

**그런데 새 CI 런(commit 8c56081124, run 31959938902)에서 `Meta Guards` 잡이 확정 `fail`로 끝났습니다.** 로그를 직접 받아 확인:
```
=== Scan: new deterministic-boundary ratchet ===
FAIL: new deterministic-boundary debt added:
origin/main...HEAD: lib/runtime/runtime_claude_code.ml:740: Option.value (int_field "cache_creation_input_tokens") ~default:0
origin/main...HEAD: lib/runtime/runtime_claude_code.ml:742: Option.value (int_field "cache_read_input_tokens") ~default:0
=== DET/NDT gate: FAIL ===
##[error]Process completed with exit code 1.
```
`scripts/ci/check-determinism-contract.sh`를 읽어보니 이 게이트는 새로 추가된 "unknown 입력 → 편리한 기본값" 패턴을 라인 단위로 잡고, 근처(`line_no-2 ~ line_no+2`, 스크립트 123-126행) 주석에 `DET-OK`/`NDT-OK`/`determinism-contract: allow`/`sound-partial: allow` 토큰이 있어야 통과시킵니다. 지금 커밋의 주석("The CLI mirrors Anthropic Messages usage... an absent cache field reads as 0, the same convention every Anthropic-format parse in agent_core uses")은 근거로는 충분하지만 **토큰이 없어서 게이트가 기계적으로 FAIL**합니다. `@check`가 초록이라는 로컬 검증은 이 게이트를 커버하지 않는 별도 스크립트라 안 잡힌 것으로 보입니다.

**판정**: 코드 수정 자체는 맞습니다. 다만 지금 상태로는 required check가 실제로 빨간불이라 "해결됨"으로 볼 수 없습니다. `DET-OK: <근거>` 형태 한 줄을 740/742행 근처(±2줄)에 추가하고 재푸시가 필요합니다.

---

### P1-2 turn_record 반박 — **REJECTED (trace_id는 재시작에서 안 바뀝니다)**

**반증 시도**: "배포(=재시작) 이후 구 record는 trace_id 가드로 구조적으로 소비 불가"라는 반박의 핵심 전제 — *재시작이 trace_id를 바꾸는가* — 를 직접 추적.

**결과**: 세 곳에서 반대 증거를 찾았습니다.
1. `lib/keeper/keeper_meta_json_parse.ml:396-397`: 영속 meta JSON을 로드할 때 `trace_id`는 `string_field fields "trace_id"`로 **그대로 읽어옵니다** — 재시작 시 재발급하는 로직이 없습니다.
2. `lib/keeper_registry/keeper_state_machine.ml:127-135`의 `Fiber_started` 처리 주석: "A new fiber = a new life... **Restart = "bring this keeper back"**" — 이 프로젝트가 스스로 "재시작"이라 부르는 이벤트(`Fiber_started`)의 리듀서는 health/buffer/backoff/stop 조건만 리셋하고 trace_id는 건드리지 않습니다.
3. trace_id가 바뀌는 유일한 이벤트는 `Handoff_completed of { new_trace_id : string; generation : int }`뿐입니다 — 이건 "재시작"이 아니라 **명시적 승계(identity reseed)**입니다. `context_fields`의 trace_id 가드 자체 주석("A reseeded identity starts a new trace")도 정확히 이 승계 케이스를 겨냥한 것이지 배포/재시작 케이스가 아닙니다.

즉 일반적인 배포(바이너리 재시작, 기존 keeper가 같은 세션/trace를 이어감)는 `Fiber_started`류 이벤트이지 `Handoff_completed`가 아닙니다. 이 경우 재시작 전후로 `meta.runtime.trace_id`가 동일하므로, 배포 직전에 기록된 구 turn_record 행의 `trace_id`도 재시작 후 `current_trace_id`와 일치 — 가드를 그대로 통과합니다. "배포=재시작"과 "trace 재발급"을 같은 것으로 놓은 전제가 코드로 반증됩니다.

**판정**: 반박 기각. turn_record는 여전히 무방비입니다 (자기치유는 "다음 턴이 기록될 때까지"만 — 승계가 아니라 재시작만으로는 그 전에 최소 1회 관측이 오염됩니다). 제가 놓친 별도의 trace_id 재발급 경로가 있다면 구체적 파일:라인으로 반증해 주십시오 — 이 부분은 제가 틀렸을 가능성도 열어둡니다만, 세 군데 독립 증거가 같은 방향을 가리켜서 확신도는 높습니다.

---

### P1-2 checkpoint `total_*` 반박 — **ACCEPTED**

**반증 시도**: "소비자 전수가 관측 표면, 게이트 0건"이라는 주장을 직접 재검증 — `total_input_tokens`에 대한 조건문/캡/스레숄드를 전수 검색.

**결과**: `checked_add`(keeper_owner_reducer.ml:244-247)는 오버플로 방지용 순수 산술이고, `nonnegative_difference`(252-255)는 단조성 sanity check일 뿐 exclusive/inclusive 혼합 자체로는 위반되지 않습니다. 유일하게 찾은 조건부 사용(`model_inference_metrics_json.ml:256`, `> 0`)은 리포트에 표시할 모델을 고르는 표시 필터라 값의 정확한 크기에 의존하는 결정이 아닙니다. 행동을 바꾸는 게이트는 발견하지 못했습니다.

**판정**: 반박 수용. 하드컷 미적용 + epoch 주석 문서화가 비례적인 처분으로 보입니다.

---

### P1-2 execution 저널 반박 — **잠정 ACCEPTED**

**반증 시도**: "소비자는 execution_event.ml 단일"인지 확인.

**결과**: `execution_event.ml:253`에서 `module Response_snapshot = Execution_provider_response_snapshot`로 구조적으로 유일 소비자임을 확인했습니다. 다만 `execution_event.ml`이 저장된 스냅샷을 재해석해서 비용/판단에 쓰는지까지는 전수 추적하지 못했습니다(시간 대비 리스크가 가장 낮다고 판단해 우선순위를 낮췄습니다).

**판정**: 구조적 근거는 확인됨. ledger 원칙(이 프로젝트의 `projects.md` "Ledger는 사실을 기록하는 evidence authority" 원칙)과도 부합해 수용하나, 확신도는 checkpoint 항목보다 낮습니다.

---

### nit (#28903 상호작용) — 수용 확인, 추가 검증 불요.

---

### 로컬 검증 claim에 대한 한계 고지

이번 라운드는 워크트리를 새 커밋으로 갱신할 권한이 없어 `@check`/`runtest`를 8c56081124 기준으로 직접 재실행하지 못했습니다. 대신 CI를 1차 증거로 썼고, 그 결과 위 Meta Guards FAIL을 발견했습니다 — 로컬 `@check` 그린이라는 보고와 실제 CI 결과가 다른 이유는 서로 다른 스크립트(빌드 vs 별도 ratchet 게이트)를 보고 있기 때문으로 보입니다. `admission 5`(50ms `admission_timeout_s`) 실패 주장은 구조적으로 타이밍 테스트가 맞아 pre-existing flake라는 설명이 그럴듯하지만, origin/main 대비 교차 실행으로 직접 확인하지는 못했습니다.

---

## 재검증 종합: **아직 PASS 아님 — 1 blocking(P0), 1 open(P1)**

- **P0 (신규, 기계적으로 확정)**: `Meta Guards` / DET-NDT ratchet이 8c56081124에서 FAIL 중입니다 (`lib/runtime/runtime_claude_code.ml:740,742`). `DET-OK: <rationale>` 주석을 해당 라인 ±2줄 안에 추가하고 재푸시해야 합니다. 이건 판단이 아니라 CI 로그로 확정된 사실입니다.
- **P1 (유지)**: turn_record 반박이 코드 근거로 성립하지 않습니다 — trace_id는 identity handoff에만 바뀌고 일반 재시작(배포)에는 유지되므로, 배포 경계를 넘어 재개되는 keeper의 최근 turn_record 행은 여전히 가드를 통과해 구 의미론을 노출합니다. 실제 하드컷(또는 다른 근거)이 더 필요합니다.
- **PASS로 인정**: P1-2의 checkpoint·execution-저널 반박은 제 독립 검증을 통과했습니다.

CI 런(31959938902)의 `Build and Test`는 제가 확인한 시점까지 여전히 pending이었습니다 — 완료되면 결과를 다시 봐 주십시오.

---
🤖 재검증 (읽기 전용, Claude 서브에이전트 자동 생성)

--- a/docs/evidence/usage-semantics-audit-2026-08-17.md
+++ b/docs/evidence/usage-semantics-audit-2026-08-17.md
@@ -43,7 +43,7 @@ Anthropic 형식 + 캐시 활성(system/tools cache_control) 조합에서:
 | 표면 | 처분 | 근거 |
 |---|---|---|
 | `cache.ml` 응답 캐시 | **schema "2" 하드컷** (이 PR) | 유일하게 구 값이 *canonical 자리로 재유입*되는 코덱 — 버전 가드가 은퇴 |
-| turn_record (RFC-0233 점유율 원천) | 조치 불요 | 점유율 소비자는 `keeper_context_observation_projection.ml`의 trace_id 가드가 현재 턴 record만 통과시킴 — 배포(=재시작) 이후 구 record는 구조적으로 소비 불가. `last_turn_usage` 표시 표면은 첫 턴에 세척되는 1턴 잔상 |
+| turn_record (RFC-0233 점유율 원천) | 조치 불요 — 단 잔류는 1턴 | **정정(재검증 반영)**: 재시작은 trace_id를 보존하므로(메타 로드가 그대로 파싱, 재발급은 Handoff뿐) trace_id 가드는 배포 경계를 거르지 못한다 — 초판의 "혼합 창이 빔" 주장은 오류. 올바른 근거: (1) record는 내부 일관 — tokens와 window가 같은 요청에서 나와 record 단위 교차-의미론 산술이 없음, 구 record의 과소 표시는 그 record가 기록될 당시 동작 그대로, (2) 다음 턴 완료가 포함형 record로 덮어써 keeper당 최대 1턴 잔류, (3) 소비자는 관측 표면(대시보드 점유율·last_turn_usage)이고 게이트 없음. 버전 필드 추가는 legacy-detection 필드라 부적격 |
 | checkpoint `usage_stats.total_*` 누적치 | 문서화(아래 epoch 주석) | 이 총합은 PR 이전부터 provider 혼합 의미론(배제형 Anthropic + 포함형 나머지)의 합이라 "순수한 v1 의미"가 애초에 없음. 전 소비자(codec 왕복, durable_event carry, tool_agent_timeline rollup, dashboard JSON ×3)가 관측 표면이고 게이트 0건. checkpoint 폐기는 transcript(durable truth) 파괴라 비례성 위반 |
 | `execution_provider_response_snapshot` 실행 저널 | 조치 불요 | 소비자는 `execution_event.ml` 단일 — 기록 시점 사실의 evidence 저널이며, 재해석/재작성은 ledger 원칙 위반 |
 

--- a/docs/evidence/usage-semantics-audit-2026-08-17.md
+++ b/docs/evidence/usage-semantics-audit-2026-08-17.md
@@ -43,7 +43,7 @@ Anthropic 형식 + 캐시 활성(system/tools cache_control) 조합에서:
 | 표면 | 처분 | 근거 |
 |---|---|---|
 | `cache.ml` 응답 캐시 | **schema "2" 하드컷** (이 PR) | 유일하게 구 값이 *canonical 자리로 재유입*되는 코덱 — 버전 가드가 은퇴 |
-| turn_record (RFC-0233 점유율 원천) | 조치 불요 — 단 잔류는 1턴 | **정정(재검증 반영)**: 재시작은 trace_id를 보존하므로(메타 로드가 그대로 파싱, 재발급은 Handoff뿐) trace_id 가드는 배포 경계를 거르지 못한다 — 초판의 "혼합 창이 빔" 주장은 오류. 올바른 근거: (1) record는 내부 일관 — tokens와 window가 같은 요청에서 나와 record 단위 교차-의미론 산술이 없음, 구 record의 과소 표시는 그 record가 기록될 당시 동작 그대로, (2) 다음 턴 완료가 포함형 record로 덮어써 keeper당 최대 1턴 잔류, (3) 소비자는 관측 표면(대시보드 점유율·last_turn_usage)이고 게이트 없음. 버전 필드 추가는 legacy-detection 필드라 부적격 |
+| turn_record (RFC-0233 점유율 원천) | 조치 불요 — 단 잔류는 1턴 | **정정(재검증 반영)**: 재시작은 trace_id를 보존하므로(메타 로드가 그대로 파싱, 재발급은 Handoff뿐) trace_id 가드는 배포 경계를 거르지 못한다 — 초판의 "혼합 창이 빔" 주장은 오류. 올바른 근거: (1) record는 내부 일관 — tokens와 window가 같은 요청에서 나와 record 단위 교차-의미론 산술이 없음, 구 record의 과소 표시는 그 record가 기록될 당시 동작 그대로, (2) 다음 턴 완료가 포함형 record로 덮어써 정상 기록 경로 기준 keeper당 최대 1턴 잔류(writer의 append는 warn-only라 기록 실패 시 상한이 늘어질 수 있음 — 기존 writer 특성), (3) 소비자는 관측 표면(대시보드 점유율·last_turn_usage)이고 게이트 없음. 버전 필드 추가는 legacy-detection 필드라 부적격 |
 | checkpoint `usage_stats.total_*` 누적치 | 문서화(아래 epoch 주석) | 이 총합은 PR 이전부터 provider 혼합 의미론(배제형 Anthropic + 포함형 나머지)의 합이라 "순수한 v1 의미"가 애초에 없음. 전 소비자(codec 왕복, durable_event carry, tool_agent_timeline rollup, dashboard JSON ×3)가 관측 표면이고 게이트 0건. checkpoint 폐기는 transcript(durable truth) 파괴라 비례성 위반 |
 | `execution_provider_response_snapshot` 실행 저널 | 조치 불요 | 소비자는 `execution_event.ml` 단일 — 기록 시점 사실의 evidence 저널이며, 재해석/재작성은 ledger 원칙 위반 |
 

--- a/lib/runtime/runtime_claude_code.ml
+++ b/lib/runtime/runtime_claude_code.ml
@@ -737,8 +737,10 @@ let parse_result ~expected_session_id ~rate_limit ~tool_effect_attempted
              { input_tokens
              ; output_tokens
              ; cache_creation_input_tokens =
+                 (* DET-OK: absent wire field is 0 by api_usage convention *)
                  Option.value (int_field "cache_creation_input_tokens") ~default:0
              ; cache_read_input_tokens =
+                 (* DET-OK: absent wire field is 0 by api_usage convention *)
                  Option.value (int_field "cache_read_input_tokens") ~default:0
              }
          | Some _, None | None, Some _ | None, None -> None)

